### PR TITLE
Added support for Windows Subsytem for Linux

### PR DESCRIPTION
**Pull request by hzahradnik**
Added support for WSL to use this extension on Windows while building on Linux. Solving the problem of translating the URI from windows filesystem to linux filesystem in #335 by defining uriConverter on the client options.
Additionally changed from defining a path to the executable for WLS to a boolean flag for WLS and instead always calling "bash.exe" with arguments "-i -c <command".

(I tried to rebase the original PR from @yisonPylkita, but failed catastrophically, because the code base changed a lot. Then I decided to start from scratch, because I changed quite a lot. I hope this is OK...)

### CODE REVIEW

**Comment by yisonPylkita** [comment]
Great ^_^

I am realy glad that you found this `uriConverter`. Can I do a code review for you ?

**Comment by hzahradnik** [comment]
Sure, knock yourself out. Thanks. :)

**Comment by yisonPylkita** [comment]
I ran a few simple tests to confirm this PR works as expected. Looks good so far - autocompletion works fine for this piece of code - **https://doc.rust-lang.org/rust-by-example/custom_types/structs.html**

Tested on Windows 10 1809, Debian as a WSL distro

**Comment by hzahradnik** [comment]
It took longer than I expected to look over your suggestions. I commited them, but unfortunately there is now a conflict. Frankly, I currently don't have the time to look into this.

**Comment by yisonPylkita** [comment]
I can resolve them for you. Just give me a write access to this PR.

This can be helpful - https://help.github.com/articles/inviting-collaborators-to-a-personal-repository/

**Comment by yisonPylkita** [comment]
@hzahradnik Thank you for write access. I will rebase this PR
@Xanewok I can make changes to this PR now. I'll look into it this weekend.

**Comment by yisonPylkita** [comment]
This rebase is still in progress. I'm on a MacBook right now. I'll check it once I return home.

File: src/utils/child_process.ts [hunk 1]
```diff
@@ -22,42 +22,59 @@ export interface ExecChildProcessResult<TOut = string> {
...
             resolve({
                 stdout,
                 stderr,
             });
         });
     });
-    return r;
 }
 
+export async function run_rustup(rustup: string, args: string[], options: child_process.ExecFileOptions, useWSL?: boolean): Promise<ExecChildProcessResult> {
```
[the remark this answers is not in the document]
Comment (yisonPylkita) [reply src/utils/child_process.ts]: I will try to move all 'run X process' related stuff to one place

File: src/extension.ts [hunk 2]
```diff
@@ -28,11 +28,13 @@ import {
...
     ServerOptions, ImplementationRequest
 } from 'vscode-languageclient';
-import { execFile, ExecChildProcessResult } from './utils/child_process';
+import { execFile, ExecChildProcessResult, run_rustup, run_rustup_process } from './utils/child_process';
+
 
 export async function activate(context: ExtensionContext) {
     configureLanguage(context);
 
+
```
Comment (Xanewok) [inline src/extension.ts]: stray newline

File: src/utils/child_process.ts [hunk 3]
```diff
@@ -22,42 +22,58 @@ export interface ExecChildProcessResult<TOut = string> {
...
-            resolve({
-                stdout,
-                stderr,
-            });
-        });
+    return execFile(command, args, options);
+}
 
-    });
+export function run_rustup_sync(rustup: string, args: string[], options: child_process.ExecFileOptions, useWSL?: boolean): Buffer {
```
Comment (Xanewok) [inline src/utils/child_process.ts]: Could we please use camelCase and move running rustup to rustup.ts file?

File: src/extension.ts (hunk 2 above)
Comment (yisonPylkita) [reply src/extension.ts]: Thanks for finding this.
Fixed

File: src/utils/child_process.ts (hunk 3 above)
Comment (yisonPylkita) [reply src/utils/child_process.ts]: Changed snake_case to camelCase. Also moved code related to running a `rustup` to rustup.ts

**Comment by yisonPylkita** [comment]
Now it looks much better.
@Xanewok Can you look at this ?

**Comment by yisonPylkita** [comment]
Did some tests on Linux. Seems fine.
Only tests on Windows w/o WSL installed left

**Comment by dtysky** [comment]
Hi @Xanewok , @yisonPylkita .

I have tested and found some problem:

1. For avoiding compile error, the "engine" filed in package.json should be "^1.26.0".

After fix this I compiled it successfully and ran debugger, then it looks like working with following output:

>C:\Applications\VSCode\Code.exe --debugBrkPluginHost=12085 --debugId=444598f3-f764-421b-8228-273f516b1f07 --extensionDevelopmentPath=C:\Users\dtysky\ComplexMind\rls-vscode 
Detected active channel: nightly-2019-01-01-x86_64-unknown-linux-gnu (since 'rust-client.channel' is unspecified)
rustup.js:231
Setting sysroot to /home/dtysky/.rustup/toolchains/nightly-2019-01-01-x86_64-unknown-linux-gnu
extension.js:348
running with rustup: rls
extension.js:376
Detected active channel: nightly-2019-01-01-x86_64-unknown-linux-gnu (since 'rust-client.channel' is unspecified)
rustup.js:231
will use a rustup-installed RLS; ensuring present

But... it is not working at all, I can not feel any features works like "auto-completion", "jump to declaration".

Is it because of the rls i slow in wsl ?

**Comment by yisonPylkita** [comment]
Hi @dtysky!
Thanks for starting tests.

Well, that is strange. I compiled this code without any warnings/errors on my setup.

`npm install` gave me this

```
➜  rls-vscode git:(support-wsl) npm install

> rust@0.5.3 postinstall /mnt/c/H_DEV/rls-vscode
> node ./node_modules/vscode/bin/install

Detected VS Code engine version: ^1.24.0
Found minimal version that qualifies engine range: 1.24.0
Fetching vscode.d.ts from: https://raw.githubusercontent.com/Microsoft/vscode/6a6e02cef0f2122ee1469765b704faf5d0e0d859/src/vs/vscode.d.ts
vscode.d.ts successfully installed!

added 320 packages from 925 contributors and audited 999 packages in 7.322s
found 1 moderate severity vulnerability
  run `npm audit fix` to fix them, or `npm audit` for details
```

and `npm run compile`

```
➜  rls-vscode git:(support-wsl) npm run compile

> rust@0.5.3 compile /mnt/c/H_DEV/rls-vscode
> tsc -p ./ 
```

---

To test this feature you have to:

1) open WSL console in Windows
2) make sure there is no rust toolchains installed (easiest way to do this is to run `rm -rf ~/.rustup ~/.cargo`. WARNING!!! This will remove all rust related stuff from your WSL. You can also move this folders somewhere else if you want to)
3) Install Rustup with nightly toolchain (In WSL console run `curl https://sh.rustup.rs -sSf | sh -s -- --default-toolchain nightly`)
4) Run this extension in `vscode` via `F5`
5) There should show up a window with new `vscode` instance. Open simple rust project there but **do not open any rust related files yet**
5) In this new `vscode` instance open File -> Preferences -> Settings -> Workspace Settings and click `{}` symbol in top right corner
6) To test this feature your configuration should be 
```json
{
    "rust-client.useWSL": true
}
```
7) Now we can check if things work as expected. In `vscode` open a rust file from your project.
8) There should show up toast 

image

Click 'Yes'
9) Wait a few seconds to give extension some time to fetch & install `rls`
10) Start typing some rust code. Code completion should work.

image


On my setup I got this from `DEBUG CONSOLE` in `vscode`

```
Detected active channel: nightly-x86_64-unknown-linux-gnu (since 'rust-client.channel' is unspecified)
Setting sysroot to /home/arlen/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu
running with rustup: rls
Detected active channel: nightly-x86_64-unknown-linux-gnu (since 'rust-client.channel' is unspecified)
will use a rustup-installed RLS; ensuring present

info: downloading component 'rust-analysis'
info: installing component 'rust-analysis'


info: downloading component 'rust-src'
info: installing component 'rust-src'

install rls

info: downloading component 'rls'
info: installing component 'rls'
```

@dtysky Can you reproduce this steps ?

**Comment by dtysky** [comment]
Hi, @yisonPylkita.

In engine 'v1.24.0', a compile error `Namespace ''vscode'' has no exported member 'DocumentSymbol'.'` will be thrown. 

Emm...After trying your given steps twice, it works. But the language analysis is too sloooooowly! (Every change will take over 10 seconds to analyze). It is far beyond my RMBP, I can't understand because my PC with windows is  I7-9700K which performance is much better than my RMBP. It is a problem of wsl ?

**Comment by Xanewok** [comment]
@dtysky what Rust project are you working on?

**Comment by dtysky** [comment]
@Xanewok

A new project created by `cargo init`.

**Comment by yisonPylkita** [comment]
Well... I didn't test a performance of this solution. Now that I think about this it should be one of the first things to do.

I will investigate this once I return home.
@dtysky Thanks for pointing this out.

**Comment by dtysky** [comment]
@yisonPylkita 

Waiting for your good job, contact me if I can do anything for this feature.

**Comment by yisonPylkita** [comment]
I did some profiling. Changed this code 
```js
if (this.config.useWSL) {
    clientOptions.uriConverters = {
        code2Protocol: (uri: Uri) => Uri.file(child_process.execFileSync('bash.exe', ['-i', '-c', `wslpath '${uri.fsPath}'`], {}).toString().trim()).toString(),
        protocol2Code: (wslPath: string) => Uri.file(child_process.execFileSync('bash.exe', ['-i', '-c', `wslpath -w ${Uri.parse(wslPath).path}`], {}).toString().trim())
    };
}
```

to this one
```js
if (this.config.useWSL) {
    clientOptions.uriConverters = {
        code2Protocol: (uri: Uri) => {
            let start = Math.floor(Date.now());
            let mapped_uri = Uri.file(child_process.execFileSync('bash.exe', ['-i', '-c', `wslpath '${uri.fsPath}'`], {}).toString().trim()).toString();
            let end = Math.floor(Date.now());
            let time_spent_in_ms = end - start;
            console.log(`code2Protocol ${time_spent_in_ms}ms`);
            return mapped_uri;
        },
        protocol2Code: (wslPath: string) => {
            let start = Math.floor(Date.now());
            let mapped_uri = Uri.file(child_process.execFileSync('bash.exe', ['-i', '-c', `wslpath -w ${Uri.parse(wslPath).path}`], {}).toString().trim());
            let end = Math.floor(Date.now());
            let time_spent_in_ms = end - start;
            console.log(`protocol2Code ${time_spent_in_ms}ms`);
            return mapped_uri;
        }
    };
}
```

and got this results

```
code2Protocol 104ms
code2Protocol 85ms
protocol2Code 76ms
code2Protocol 104ms
code2Protocol 95ms
code2Protocol 100ms
code2Protocol 86ms
code2Protocol 84ms
code2Protocol 81ms
code2Protocol 254ms
code2Protocol 177ms
code2Protocol 87ms
code2Protocol 90ms
code2Protocol 86ms
code2Protocol 85ms
code2Protocol 93ms
code2Protocol 102ms
protocol2Code 86ms
protocol2Code 92ms
protocol2Code 84ms
```

Around **90ms** is a lot of time and this routine is called quite often (pretty much every a few key strokes). I will replace spawning `wslpath` process with 2 functions - `uriWslToWindows` and `uriWindowsToWsl` and then run the benchmark once again.

Yet this will have to wait for tomorrow. Once I am done with this I will let you @dtysky know.

**Comment by Xanewok** [comment]
@yisonPylkita did you come back to that, by any chance?

**Comment by yisonPylkita** [comment]
I started working on path translation in https://github.com/yisonPylkita/wslpath.ts

I already implemented simple paths translation from WSL to Windows. Still need some time to finish path translation from Windows to WSL

**Comment by yisonPylkita** [comment]
@dtysky Can you check this now ?

**Comment by dtysky** [comment]
@yisonPylkita 

How do I test it? Need any new options?

**Comment by yisonPylkita** [comment]
It should work right away. Just open a simple Rust project with workspace configuration the same as before.

I am most interested in performance. Is this finally usable for you ? Now path translation takes less than **1ms** instead of **~90ms** like before.

**Comment by dtysky** [comment]
@yisonPylkita Every changes still took about 20s to to analyze...

**Comment by yisonPylkita** [comment]
WSL can be sometimes really slow, especially when it comes to I/O operations. [Link](https://www.reddit.com/r/bashonubuntuonwindows/comments/a23nul/very_slow_performance_is_this_normal_for_wsl/) Maybe `rls` itself runs very slowly on your WSL instance and it takes **20ms* for it to send data about autocompletion.

Give me more time. I will try to reproduce your problem.
Can you send me some info about your system ?
I will need:
- Source snippet you are using for tests
- Your Windows version (1709, 1803, or maybe 1809)
- Do you have an antivirus software installed ? Do you have enabled Windows Defender ?
- Please run `Measure-Command { wsl ls }` in your PowerShell console. It will check how long it took to run process with I/O operations on WSL

**Comment by dtysky** [comment]
@yisonPylkita

Sorry for late reply, I have too much works this week.

1. Source snippet you are using for tests: Just a new project has single entry file:

```rs
fn main() {
    println!("Hello, world!");
    let a = 1;
    a = x;
}
```

2. Your Windows version (1709, 1803, or maybe 1809): 1809
3. Do you have an antivirus software installed ? Do you have enabled Windows Defender ?: I use Windows Defender and has disabled it.
4. Please run `Measure-Command { wsl ls }` in your PowerShell console. It will check how long it took to run process with I/O operations on WSL:

```txt
Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 113
Ticks             : 1131024
TotalDays         : 1.30905555555556E-06
TotalHours        : 3.14173333333333E-05
TotalMinutes      : 0.00188504
TotalSeconds      : 0.1131024
TotalMilliseconds : 113.1024
```

**Comment by SambeauWang** [comment]
@yisonPylkita I follew your suggestion: build and install this extension. But I meet the follew the question.
`
Error getting sysroot from `rustc`: Error: Command failed: wsl rustup run nightly rustc --print sysroot
extension.js:337
/bin/bash: rustup: command not found
Let's retry with extended $PATH
extension.js:338
Error reading sysroot (second try) Error: Error getting sysroot from `rustc`: Error: Command failed: wsl rustup run nightly rustc --print sysroot
extension.js:344
/bin/bash: rustup: command not found
	at ClientWorkspace.<anonymous> (D:\Temp\rls\rls-vscode\out\src\extension.js:319:23)
	at Generator.throw (<anonymous>)
	at rejected (D:\Temp\rls\rls-vscode\out\src\extension.js:14:65)
running with rustup: rls
extension.js:377
Error: Command failed: wsl rustup toolchain list
rustup.js:99
/bin/bash: rustup: command not found
	at ChildProcess.exithandler (child_process.js:303:12)
	at ChildProcess.emit (events.js:182:13)
	at maybeClose (internal/child_process.js:961:16)
	at Process.ChildProcess._handle.onexit (internal/child_process.js:248:5)
rejected promise not handled within 1 second
`
Then,  I try to find the solution. Because wsl.exe never load .bashrc file, when I run the command. how to deal with it?

**Comment by yisonPylkita** [comment]
I also have Windows 10 1809

On my setup command `Measure-Command { wsl ls }` took

```
Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 43
Ticks             : 438022
TotalDays         : 5.06969907407407E-07
TotalHours        : 1.21672777777778E-05
TotalMinutes      : 0.000730036666666667
TotalSeconds      : 0.0438022
TotalMilliseconds : 43.8022
```

Difference between yours and mine result isn't that great. Yours code completion should work fine too.

Try with this `vscode` configuration

```json
{
    "rust-client.useWSL": true,
    "rust-client.rustupPath": "~/.cargo/bin/rustup"
}
```

For me it works like this

deepbeneficialbullfrog

**Comment by SambeauWang** [comment]
When i goto the test function definition in main.rs, I get the error：Unable to open '': File is a directory.
```
use std::io;
use rand::Rng;

fn test(){
    println!("test")
}

fn main() {
    let secret_number = rand::thread_rng().gen_range(1, 101);
    println!("The Secret Number: {}", secret_number);

    println!("Guess the num:");
    let mut guess = String::new();
    io::stdin().read_line(&mut guess)
        .expect("Failed to read line.");
    println!("You guessed: {}", guess);

    test();
}
```

**Comment by dtysky** [comment]
@yisonPylkita

Still very slow, my wsl system is ubuntu 16.04, and versions of my rust are:

rustc 1.34.0-nightly (b139669f3 2019-02-06)
rustc 1.34.0-nightly (b139669f3 2019-02-06)  
rls 1.34.0 (f331ff7 2019-02-03)

And here is recently result of `Measure-Command { wsl ls }`:

```txt
Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 33
Ticks             : 331591
TotalDays         : 3.8378587962963E-07
TotalHours        : 9.21086111111111E-06
TotalMinutes      : 0.000552651666666667
TotalSeconds      : 0.0331591
TotalMilliseconds : 33.1591
```

My settings of vscode:

```
{
    "rust-client.useWSL": true,
    "rust-client.rustupPath": "~/.cargo/bin/rustup"
}
```

What should I do to fix this problem ?

**Comment by yisonPylkita** [comment]
@SambeauWang Thanks for your participation. I will look into this issue.

Do you have long waiting times for autocompletion ?

**Comment by yisonPylkita** [comment]
@dtysky This is problematic because I cannot reproduce your problem. But maybe it is because of Ubuntu version. I might be using Ubuntu 18.04 while you are using Ubuntu 16.04. I will check autocompletion on Ubuntu 16.04.

**Comment by yisonPylkita** [comment]
@dtysky I made some tests on `Ubuntu 16.04` and it is **way slower** than `Ubuntu 18.04`.

For this code any change in `Vec3d` took `rls` something around **15s**

```rust
struct Vec3d {
    x: i32,
    y: i32,
    z: i32,
}

fn main() {
    let s = "Hi there ^_^";

    for ch in s.chars() {
        println!("{}", ch);
    }

    let mut _v = Vec3d{
        x: 0, y: 0, z: 0,
    };
}
```

Can you check performance on `Ubuntu 18.04` ?

(Make sure to check if `Ubuntu 18.04` is your default WSL instance. Run `wsl lsb_release -a` to confirm this)

**Comment by dtysky** [comment]
@yisonPylkita

I have tried to upgrade my ubuntu from 16.04 to 18.04 and no better.

I will try to reinstall rustup, rustc and rls this weekend.

PS: Really don't want to reinstall ubuntu again...... I have reinstalled and configurated it twice, installing some packages like node, rust, zsh in china are very very very slow(even if using VPN) -_-

File: src/rustup.ts [hunk 4]
```diff
@@ -213,21 +218,59 @@ export function parseActiveToolchain(rustupOutput: string): string {
...
+    let command: string = rustup;
+
+    if (useWSL == true) {
+        ({ command: command, args: args } = modifyParametersForWSL(command, args));
+    }
+
+    return execFile(command, args, options);
+}
+
+export function runRustupSync(rustup: string, args: string[], options: child_process.ExecFileOptions, useWSL?: boolean): Buffer {
```
Comment (Xanewok) [inline src/rustup.ts]: Same as above
```suggestion
export function execCmdSync(command: string, args: string[], options: child_process.ExecFileOptions, useWSL?: boolean): Buffer {
```

File: src/rustup.ts [hunk 5]
```diff
@@ -213,21 +218,59 @@ export function parseActiveToolchain(rustupOutput: string): string {
...
+    let command: string = rustup;
+
+    if (useWSL == true) {
+        ({ command: command, args: args } = modifyParametersForWSL(command, args));
+    }
+
+    return child_process.execFileSync(command, args, { ...options });
+}
+
+export function runRustupProcess(rustup: string, args: string[], options: child_process.ExecFileOptions, useWSL?: boolean): child_process.ChildProcess {
```
Comment (Xanewok) [inline src/rustup.ts]: Same as above

```suggestion
export function runProcess(command: string, args: string[], options: child_process.ExecFileOptions, useWSL?: boolean): child_process.ChildProcess {
```

File: src/extension.ts [hunk 6]
```diff
@@ -30,6 +31,7 @@ import {
 } from 'vscode-languageclient';
 import { execFile, ExecChildProcessResult } from './utils/child_process';
 
+
```
Comment (Xanewok) [inline src/extension.ts]: Stray newline

File: src/configuration.ts [hunk 7]
```diff
@@ -86,8 +91,9 @@ export class RLSConfiguration {
         this.wsPath = wsPath;
     }
 
-    public rustupConfig(): RustupConfig {
-        return new RustupConfig(this.channel, this.rustupPath);
+    // Added ignoreChannel for readChannel function. Otherwise we end in an infinite loop.
```
Comment (Xanewok) [inline src/configuration.ts]: I guess it'd be better to restructure this but we can leave this for a follow up PR

File: src/utils/wslpath.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    if (uriSegments[uriSegments.length - 1] == '') {
+      uriSegments.pop();
+    }
+  
+    const disc_letter = uriSegments[1].toUpperCase();
+    if (!/^[A-Z]+$/.test(disc_letter)) {
+      return '';
+    }
+    uriSegments.shift(); // remove mnt
+    uriSegments.shift(); // remove disc letter
```
Comment (Xanewok) [inline src/utils/wslpath.ts]: nit
```suggestion
    uriSegments.shift(); // remove disk letter
```

**Comment by yisonPylkita** [comment]
@dtysky You do not have to remove your `Ubuntu 16.04` or upgrade it. There is a `Ubuntu 18.04` as standalone application.

image

WSL lets you have multiple WSL instances installed at the same time.
To select default one follow this guide - https://docs.microsoft.com/en-us/windows/wsl/wsl-config#versions-earlier-than-windows-10-version-1903

File: src/rustup.ts (hunk 5 above)
Comment (yisonPylkita) [reply src/rustup.ts]: `spawnProcess` seems better. Internally it calls `spawn()` from `Node.js` so some with that background will have an idea what that function is doing just by a name

**Comment by yisonPylkita** [comment]
@Xanewok Thanks for this review. I fixed most of it. Only one left is 
`Added ignoreChannel for readChannel function...`. I will check this later.

File: src/rustup.ts (hunk 5 above)
Comment (Xanewok) [reply src/rustup.ts]: Sounds good!

**Comment by dtysky** [comment]
@yisonPylkita

I have tried to install a new ubuntu18.04 and this plugin works great ! Thanks for your contribution !

**Comment by yisonPylkita** [comment]
Yay! I am happy to hear that ^_^

I will have to add a note about *Ubuntu 16.04* in README

**Comment by Xanewok** [comment]
Many thanks @hzahradnik and @yisonPylkita for bringing this across the finish line! From the looks of it it seems to work and I can confirm that this continues to work in non-WSL scenarios so I'll be happy to merge this :tada:

**Comment by yisonPylkita** [comment]
@SambeauWang I retested your snippet

```rust
use std::io;
use rand::Rng;

fn test(){
    println!("test")
}

fn main() {
    let secret_number = rand::thread_rng().gen_range(1, 101);
    println!("The Secret Number: {}", secret_number);

    println!("Guess the num:");
    let mut guess = String::new();
    io::stdin().read_line(&mut guess)
        .expect("Failed to read line.");
    println!("You guessed: {}", guess);

    test();
}
```

Goto `test()` function works (F12 on `test();` in main()). The thing that does not work is goto  `rand::thread_rng().gen_range()`. 

I could write a fix for this but it will make Windows application (vscode) open and possibly overwrite (on user modifications) this library file. 

One of the WSL guys made a post on Windows blog about this - https://blogs.msdn.microsoft.com/commandline/2016/11/17/do-not-change-linux-files-using-windows-apps-and-tools/

He does not recommend opening WSL files in Windows applications. 

image

So for now I will leave things as they are. Maybe in the (not so distant) future Windows will have this feature. Then of course `rls-vscode` will support it.

**Comment by Systemcluster** [comment]
@yisonPylkita This feature will be added in the coming April update: https://blogs.msdn.microsoft.com/commandline/2019/02/15/whats-new-for-wsl-in-windows-10-version-1903/

**Comment by yisonPylkita** [comment]
@Systemcluster Nice! Thank you for sharing this.
Once I get April patch (1903) I will start working on this feature.

**Comment by kobenauf** [comment]
I was trying to follow steps here, including these in vsc settings:
```
    "rust-client.useWSL": true,
    "rust-client.rustupPath": "~/.cargo/bin/rustup"
```

But I kept getting failures when it tried to install RLS server.  terminal window reported:

> The terminal process terminated with exit code: 1

And two popup dialogs opened showing:
> Couldn't start client Rust Language Server
> Could not install component: `rust-src`

I eventually realized it was because I had both `"rust-client.useWSL": true,` and also this setting:
```
"terminal.integrated.shell.windows": "C:\\WINDOWS\\System32\\wsl.exe",
```
When I removed `terminal.integrated.shell.windows` it was able to install the RLS server.

**Pull request merged**